hazelcast/src/main/java/com/hazelcast/config/DynamicClassloadingConfig.java

```diff
@@ -0,0 +1,239 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.config;
+
+import com.hazelcast.spi.annotation.Beta;
+
+/**
+ * Configuration of dynamic classloading. When enabled it allows members to load classes from other members.
+ * This simplifies deployment as you do not have to deploy your domain classes into classpath of all cluster members.
+ *
+ */
+
+@Beta
+public class DynamicClassloadingConfig {
+
+    /**
+     * Controls caching of caches loaded from remote members
+     */
+    public enum ClassCacheMode {
+        /**
+         * Never caches loaded classes. This is suitable for loading runnables, callables, entry processors, etc.
+         *
+         */
+        OFF,
+
+        /**
+         * Cache indefinitely. This is suitable when you load long-living objects, such as domain objects stored
+         * in a map.
+         *
+         */
+        ETERNAL,
+    }
+
+    /**
+     * Controls how to react on receiving a classloading request from a remote member
+     *
+     */
+    public enum ProviderMode {
+        /**
+         * Never serves classes to other members. This member will never server classes to remote members.
+          */
+        OFF,
+
+        /**
+         * Serve classes from local classpath only. Classes loaded from other members will be used locally,
+         * but they won't be served to other members.
+         *
+         */
+        LOCAL_CLASSES_ONLY,
+
+        /**
+         * Serve classes loaded from both local classpath and from other members.
+         *
+         */
+        LOCAL_AND_CACHED_CLASSES
+    }
+
+    private ClassCacheMode classCacheMode = ClassCacheMode.ETERNAL;
+    private ProviderMode providerMode = ProviderMode.LOCAL_AND_CACHED_CLASSES;
+    private String blacklistedPrefixes;
+    private String whitelistedPrefixes;
+    private String providerFilter;
+    private boolean enabled;
+
+    /**
+     * Enable or disable dynamic classloading. Default: {{@code false}}
+     *
+     * @param enabled
+     * @return
+     */
+    public DynamicClassloadingConfig setEnabled(boolean enabled) {
+        this.enabled = enabled;
+        return this;
+    }
+
+    /**
+     * Returns true when dynamic classloading is enabled
+     *
+     * @return {{@code true}} when dynamic classloading is enabled
+     */
+    public boolean isEnabled() {
+        return enabled;
+    }
+
+    /**
+     * Filter to constraint members to be used for classloading request when a class
+     * is not available locally.
+     *
+     * Filter format: <code>HAS_ATTRIBUTE:foo</code> this will send classloading requests
```

Comment:
more filters are definitely an option. I was even considering a custom filter impl. `CUSTOM:com.bar.FooFilter`, but not for the 1st version. 

I'll add a better example, that's a very good hint.